File: backend/routes/bookings.py
from flask import Blueprint, request, jsonify
from sqlalchemy.exc import IntegrityError
from werkzeug.exceptions import BadRequest, Conflict
import logging
from datetime import datetime
from backend.models.booking import Booking
from backend.middleware.auth import login_required
from backend.config.database import db_session

# Setup logging
logging.basicConfig(level=logging.INFO,
                    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

booking_bp = Blueprint('bookings', __name__)

@booking_bp.route('/api/bookings', methods=['POST'])
@login_required
def create_booking(current_user):
    """
    Create a new booking with uniqueness constraint validation
    """
    try:
        logger.info(f"Processing booking request for user: {current_user['id']}")
        data = request.get_json()

        # Validate required fields
        required_fields = ['office_id', 'slot_id', 'booking_date']
        for field in required_fields:
            if field not in data:
                logger.error(f"Missing required field: {field}")
                return jsonify({'error': f"Missing required field: {field}"}), 400

        # Parse booking date
        try:
            booking_date = datetime.fromisoformat(data['booking_date'].replace('Z', '+00:00'))
        except ValueError as e:
            logger.error(f"Invalid date format: {e}")
            return jsonify({'error': f"Invalid date format: {str(e)}"}), 400

        # Create new booking
        new_booking = Booking(
            user_id=current_user['id'],
            office_id=data['office_id'],
            slot_id=data['slot_id'],
            booking_date=booking_date,
            status="confirmed"
        )

        # Save booking with uniqueness constraint
        try:
            db_session.add(new_booking)
            db_session.commit()
            logger.info(f"Booking created successfully: {new_booking.id}")
            return jsonify(new_booking.to_dict()), 201
        except IntegrityError as e:
            db_session.rollback()
            if "unique_booking" in str(e):
                logger.warning(f"Duplicate booking attempted: {str(e)}")
                return jsonify({'error': 'You already have a booking for this slot and date'}), 409
            else:
                logger.error(f"Database integrity error: {str(e)}")
                return jsonify({'error': 'Database constraint violation'}), 400

    except Exception as e:
        logger.error(f"Error creating booking: {str(e)}")
        return jsonify({'error': f"Server error: {str(e)}"}), 500

@booking_bp.route('/api/bookings', methods=['GET'])
@login_required
def get_user_bookings(current_user):
    """Get all bookings for the current user"""
    try:
        bookings = Booking.query.filter_by(user_id=current_user['id']).all()
        return jsonify([booking.to_dict() for booking in bookings])
    except Exception as e:
        logger.error(f"Error fetching bookings: {str(e)}")
        return jsonify({'error': f"Server error: {str(e)}"}), 500

# Error handlers
@booking_bp.errorhandler(BadRequest)
def handle_bad_request(error):
    return jsonify({'error': str(error)}), 400

@booking_bp.errorhandler(Conflict)
def handle_conflict(error):
    return jsonify({'error': str(error)}), 409

File: backend/routes/bookings.js
const express = require('express');
const mongoose = require('mongoose');
const router = express.Router();
const Booking = require('../models/Booking');
const auth = require('../middleware/auth');
const logger = require('../config/logger');

/**
 * @route   POST /api/bookings
 * @desc    Create a new booking with uniqueness constraint
 * @access  Private
 */
router.post('/', auth, async (req, res) => {
  const session = await mongoose.startSession();
  session.startTransaction();

  try {
    const { officeId, slotId, date } = req.body;
    const userId = req.userData.userId;

    // Input validation
    if (!officeId || !slotId || !date) {
      return res.status(400).json({
        error: 'Missing required fields: officeId, slotId, and date are required'
      });
    }

    // Validate IDs format
    if (!mongoose.Types.ObjectId.isValid(officeId) || !mongoose.Types.ObjectId.isValid(slotId)) {
      return res.status(400).json({ error: 'Invalid officeId or slotId format' });
    }

    // Validate date
    const bookingDate = new Date(date);
    if (isNaN(bookingDate.getTime())) {
      return res.status(400).json({ error: 'Invalid date format' });
    }

    // Check if date is in the past
    const today = new Date();
    today.setHours(0, 0, 0, 0);
    if (bookingDate < today) {
      return res.status(400).json({ error: 'Cannot book for a date in the past' });
    }

    // Check for duplicate booking
    const isDuplicate = await Booking.checkDuplicateBooking(userId, officeId, date);
    if (isDuplicate) {
      return res.status(409).json({
        error: 'You already have a booking for this office on this date'
      });
    }

    // Create new booking
    const newBooking = new Booking({
      userId,
      officeId,
      slotId,
      date: bookingDate,
      status: 'pending'
    });

    // Save booking
    const savedBooking = await newBooking.save({ session });
    await session.commitTransaction();

    logger.info(`New booking created: ${savedBooking._id} for user ${userId}`);
    res.status(201).json({
      message: 'Booking created successfully',
      booking: savedBooking
    });

  } catch (error) {
    await session.abortTransaction();

    // Check for duplicate key error (MongoDB error code for duplicate key is 11000)
    if (error.code === 11000) {
      logger.warn('Duplicate booking attempt:', error.message);
      return res.status(409).json({
        error: 'You already have a booking for this office on this date'
      });
    }

    logger.error('Error creating booking:', error);
    res.status(500).json({
      error: 'An error occurred while creating the booking',
      details: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  } finally {
    session.endSession();
  }
});

// Get all bookings (for testing)
router.get('/', auth, async (req, res) => {
  try {
    const bookings = await Booking.find({ userId: req.userData.userId })
      .sort({ date: -1 });
    res.json(bookings);
  } catch (error) {
    logger.error('Error fetching bookings:', error);
    res.status(500).json({ error: 'Error fetching bookings' });
  }
});

module.exports = router;
